src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldMapperImpl.java

```diff
@@ -69,24 +71,34 @@ public void handleField(Resource parent, DataField dataField, List<ControlField>
         mappedResource = addNewEdge(parentResource, dataField, controlFields, fieldRule);
         addRelation(parentResource, mappedResource, dataField, fieldRule);
       }
-      var mapper = !EMPTY.equals(dataField.getTag())
-        ? ofNullable(marc2ldMappersMap.get(dataField.getTag()))
-        : controlFields.stream()
-        .map(controlField -> marc2ldMappersMap.get(controlField.getTag()))
-        .filter(Objects::nonNull)
-        .filter(m -> m.canMap(valueOf(fieldRule.getPredicate())))
-        .findFirst();
-      mapper.ifPresent(m -> {
-        if (m.canMap(valueOf(fieldRule.getPredicate()))) {
-          m.map(new MarcData(dataField, controlFields), mappedResource);
-        }
-      });
+
+      var mapper = findMapper(dataField, controlFields, fieldRule);
```

Comment:
minor

I try to make the order of steps clearer
```java
findMapper(dataField.getTag(), fieldRule)
        .or(() -> findMapper(controlFields, fieldRule))
        .ifPresent((m -> m.map(new MarcData(dataField, controlFields), mappedResource)));
```
ann separate logic data field and control field (the is no parameters together in methods)
```java
  private Optional<Marc2ldMapper> findMapper(List<ControlField> controlFields, Marc4BibframeRules.FieldRule fieldRule) {
    return controlFields.stream()
      .map(ControlField::getTag)
      .flatMap(tag -> findMapper(tag, fieldRule).stream())
      .findFirst();
  }

  private Optional<Marc2ldMapper> findMapper(String tag, Marc4BibframeRules.FieldRule fieldRule) {
    if (StringUtils.isEmpty(tag)) {
      return Optional.empty();
    }
    return marc2ldMappersMap.getOrDefault(tag, emptyList())
      .stream()
      .filter(mapper -> mapper.canMap(valueOf(fieldRule.getPredicate())))
      .findFirst();
  }
```
bonuses
- less parameters in methods
- see order of algorithm: 1 first find in data field, then find in control field)

Reply:
Good idea. Will fix

Reply:
Done